File: src/main/java/se/kth/spork/spoon/conflict/CommentContentHandler.java
package se.kth.spork.spoon.conflict;

import java.util.Optional;
import se.kth.spork.util.LineBasedMerge;
import se.kth.spork.util.Pair;
import spoon.reflect.declaration.CtElement;
import spoon.reflect.path.CtRole;

/**
 * A conflict handler for comment contents.
 *
 * @author Simon Larsén
 */
public class CommentContentHandler implements ContentConflictHandler {
    @Override
    public CtRole getRole() {
        return CtRole.COMMENT_CONTENT;
    }

    @Override
    public Pair<Optional<Object>, Boolean> handleConflict(
            Optional<Object> baseVal,
            Object leftVal,
            Object rightVal,
            Optional<CtElement> baseElem,
            CtElement leftElem,
            CtElement rightElem) {
        return Pair.of(mergeComments(baseVal.orElse(""), leftVal, rightVal), false);
    }

    private static Optional<Object> mergeComments(Object base, Object left, Object right) {
        Pair<String, Integer> merge =
                LineBasedMerge.merge(base.toString(), left.toString(), right.toString());
        if (merge.second > 0) {
            return Optional.empty();
        }
        return Optional.of(merge.first);
    }
}

File: src/main/kotlin/se/kth/spork/spoon/conflict/CommentContentHandler.kt
package se.kth.spork.spoon.conflict

import spoon.reflect.path.CtRole
import spoon.reflect.declaration.CtElement
import se.kth.spork.util.LineBasedMerge
import se.kth.spork.util.Pair

/**
 * A conflict handler for comment contents.
 *
 * @author Simon Larsén
 */
class CommentContentHandler : ContentConflictHandler {
    override val role: CtRole
        get() = CtRole.COMMENT_CONTENT

    override fun handleConflict(
        baseVal: Any?,
        leftVal: Any,
        rightVal: Any,
        baseElem: CtElement?,
        leftElem: CtElement,
        rightElem: CtElement
    ): Pair<Any?, Boolean> {
        return Pair.of(mergeComments(baseVal ?: "", leftVal, rightVal), false)
    }

    private fun mergeComments(base: Any, left: Any, right: Any): Any? {
        val merge = LineBasedMerge.merge(base.toString(), left.toString(), right.toString())
        return if (merge.second > 0) null
        else merge.first
    }
}